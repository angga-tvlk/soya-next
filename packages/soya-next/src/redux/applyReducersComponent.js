import React from 'react';
import { storeShape } from 'react-redux/lib/utils/PropTypes';
import getDisplayName from '../utils/getDisplayName';

export default reducers => Component => {
  if (typeof reducers === 'undefined') {
    return Component;
  }

  class ApplyReducers extends React.Component {
    static displayName = getDisplayName('ApplyReducers', Component);

    static contextTypes = {
      store: storeShape.isRequired,
    };

    static propTypes = {
      store: storeShape,
    };

    constructor(props, context) {
      super(props, context);
      this.store = props.store || context.store;
      if (!this.store.soya) {
        throw new Error('applyReducers must be used with Soya\'s redux enhancer');
      }
      this.store.addReducer(reducers);
    }

    render() {
      return <Component {...this.props} />;
    }
  }
  return ApplyReducers;
};
