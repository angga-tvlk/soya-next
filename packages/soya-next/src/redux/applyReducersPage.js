import React from 'react';
import getDisplayName from '../utils/getDisplayName';

export default reducers => Page => {
  if (typeof reducers === 'undefined') {
    return Page;
  }

  class ApplyReducers extends React.Component {
    static displayName = getDisplayName('ApplyReducers', Page);

    static async getInitialProps(ctx) {
      if (!ctx.store.soya) {
        throw new Error('applyReducers must be used with Soya\'s redux enhancer');
      }
      ctx.store.addReducer(reducers);
      return Page.getInitialProps && await Page.getInitialProps(ctx);
    }

    render() {
      return <Page {...this.props} />;
    }
  }
  return ApplyReducers;
};
