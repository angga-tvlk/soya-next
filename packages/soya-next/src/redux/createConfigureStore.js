import thunk from 'redux-thunk';
import {
  applyMiddleware,
  combineReducers,
  createStore,
} from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/logOnlyInProduction';
import enhancer from './storeEnhancer';
import middleware from './middleware';

export default preloadedReducers => (preloadedState, extraArgument) => createStore(
  preloadedReducers ? combineReducers(preloadedReducers) : () => ({}),
  preloadedState,
  composeWithDevTools(
    applyMiddleware(
      thunk.withExtraArgument(extraArgument),
      middleware,
    ),
    enhancer(preloadedReducers),
  ),
);
