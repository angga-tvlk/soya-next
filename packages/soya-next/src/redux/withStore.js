import React from 'react';
import { Cookies } from 'react-cookie';
import PropTypes from 'prop-types';
import { storeShape } from 'react-redux/lib/utils/PropTypes';
import getDisplayName from '../utils/getDisplayName';

export default (configureStore, preloadedReducers) => Page => {
  const configureStoreIfNeeded = (preloadedState, extraArgument) => {
    if (typeof window === 'undefined') {
      return configureStore(preloadedReducers, preloadedState, extraArgument);
    }
    if (typeof window.store === 'undefined') {
      window.store = configureStore(preloadedReducers, preloadedState, extraArgument);
    } else {
      window.store.addReducer(preloadedReducers);
    }
    return window.store;
  };

  class WithStore extends React.Component {
    static displayName = getDisplayName('WithStore', Page);

    static propTypes = {
      cookies: PropTypes.instanceOf(Cookies).isRequired,
      preloadedState: PropTypes.object.isRequired,
      store: PropTypes.oneOfType([
        PropTypes.object,
        storeShape,
      ]).isRequired,
    };

    static async getInitialProps(ctx) {
      const store = configureStoreIfNeeded(undefined, { cookies: ctx.cookies });
      const props = Page.getInitialProps && await Page.getInitialProps({ ...ctx, store });
      return {
        ...props,
        preloadedState: store.getState(),
        store,
      };
    }

    constructor(props) {
      super(props);
      this.store = props.store.dispatch ? props.store : configureStoreIfNeeded(props.preloadedState, { cookies: props.cookies });
      if (!this.store.soya) {
        throw new Error('withStore must be used with Soya\'s redux enhancer');
      }
    }

    render() {
      const { ...props } = this.props;
      delete props.preloadedState;
      delete props.store;

      return (
        <Page {...props} store={this.store} />
      );
    }
  }

  return WithStore;
};
