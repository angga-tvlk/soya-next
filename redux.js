exports.applyReducers = require('./lib/redux/applyReducers');
exports.storeEnhancer = require('./lib/redux/storeEnhancer');
