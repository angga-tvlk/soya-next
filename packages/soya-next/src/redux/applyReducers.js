import applyReducersComponent from './applyReducersComponent';
import applyReducersPage from './applyReducersPage';

export default reducers => Component => {
  if (typeof Component.getInitialProps === 'undefined') {
    return applyReducersComponent(reducers)(Component);
  }
  return applyReducersPage(reducers)(Component);
};
