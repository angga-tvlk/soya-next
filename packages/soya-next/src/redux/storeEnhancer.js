import { combineReducers } from 'redux';

export default globalReducers => createStore => (reducer, preloadedState, enhancer) => {
  const store = createStore(reducer, preloadedState, enhancer);
  const replaceReducer = store.replaceReducer;
  let soyaReducers = { ...globalReducers };

  return {
    ...store,
    addReducer: nextReducers => {
      if (!nextReducers) {
        throw new Error('Missing nextReducers argument.');
      }

      let nextReducer = nextReducers;
      if (typeof nextReducers === 'object') {
        const keys = Object.keys(nextReducers);
        keys.forEach(key => {
          if (soyaReducers[key] && soyaReducers[key] !== nextReducers[key]) {
            throw new Error(`Duplicate reducer name: ${key}`);
          }
          soyaReducers[key] = nextReducers[key];
        });
        nextReducer = combineReducers(soyaReducers);
      }
      replaceReducer(nextReducer);
    },
    replaceReducer: nextReducers => {
      if (!nextReducers) {
        throw new Error('Missing nextReducers argument.');
      }

      let nextReducer = nextReducers;
      if (typeof nextReducers === 'object') {
        soyaReducers = {
          ...globalReducers,
          ...nextReducers,
        };
        nextReducer = combineReducers(soyaReducers);
      }
      replaceReducer(nextReducer);
    },
    soya: true,
  };
};
