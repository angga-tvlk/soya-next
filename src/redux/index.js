import storeEnhancer from './storeEnhancer';
import withReducers from './withReducers';
import withStore from './withStore';

export { storeEnhancer, withReducers, withStore };
