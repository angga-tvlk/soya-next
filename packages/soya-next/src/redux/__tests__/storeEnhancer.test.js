import { createStore } from 'redux';
import storeEnhancer from '../storeEnhancer';

describe('Store Enhancer', () => {
  let store;
  const todos = (state = [], action) => state;

  beforeEach(() => {
    store = storeEnhancer()(createStore)(state => state);
  });

  it('should expose enhanced API', () => {
    const apis = Object.keys(store);
    expect(apis.length).toBe(6);
    expect(apis).toContain('addReducer');
    expect(apis).toContain('replaceReducer');
    expect(apis).toContain('soya');
  });

  it('should add new todos to store state', () => {
    store.addReducer({ todos });
    expect(store.getState()).toMatchSnapshot();
  });

  it('should replace store state with given reducers', () => {
    const auth = (state = {}) => state;
    store.addReducer({ auth });
    store.addReducer({ todos });
    store.replaceReducer({ auth });
    expect(store.getState()).toMatchSnapshot();
  });

  it('should throw if addReducers arguments are not specified', () => {
    expect(() => {
      store.addReducer();
    })
      .toThrow('Missing nextReducers argument.');
  });

  it('should throw if replaceReducers arguments are not specified', () => {
    expect(() => {
      store.replaceReducer();
    })
      .toThrow('Missing nextReducers argument.');
  });

  it('should throw if there\'s reducer name conflict', () => {
    expect(() => {
      store.addReducer({ todos });
      store.addReducer({ todos: state => state });
    })
      .toThrow('Duplicate reducer name: todos');
  });
});
