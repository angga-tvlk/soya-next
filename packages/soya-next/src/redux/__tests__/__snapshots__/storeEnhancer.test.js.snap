// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Enhancer should add new todos to store state 1`] = `
Object {
  "todos": Array [],
}
`;

exports[`Store Enhancer should replace store state with given reducers 1`] = `
Object {
  "auth": Object {},
  "todos": Array [],
}
`;
